### ISSUE TITLE
ini file locations

### ISSUE THREAD

**Issue by pryrt** [vinsworldcom/nppQuickText#17]
I just manually installed 0.1.3.1 at home, where I have a normally-installed Notepad++.  When I don't run NPP as Administrator, I cannot add new shortcuts into `QuickText.ini` (either thru the GUI, or through editing `%ProgramFiles%\Notepad++\Plugins\QuickText\QuickText.ini`).  

Further, having one config file in `Config`, and one in the DLL directory, is confusing.  And if you _are_ going to have that confusion, it would make more sense for the one that influences the GUI (the `QuickText.conf.ini`) with the DLL, and the one that users will usually be editing (`QuickText.ini`, with all the shortcut definitions) in the Config directory.

But really, the plugin should just grab NPPM_GETPLUGINSCONFIGDIR, and put both `QuickText.ini` and `QuickText.conf.ini` in that folder -- or, better, in `<NPPM_GETPLUGINSCONFIGDIR>\QuickText\` -- that will then properly put it in a writeable location, whether it is an installed Notepad++ or portable Notepad++, and they will be together, and (with the subdir) hierarchically separated from the other plugins' config files.

**Comment by vinsworldcom** [comment]
@pryrt 
I did give that some thought as they way they were originally finding the plugin config directory was wonky.  Also, using custom parsing for an INI file instead of just the API calls (Get/WritePrivateProfile*()).  Alas, the legacy code proved harder to fix than I thought so I gave up.

As I mentioned in #16 (which you helped answer) I didn't intend for this to be installed via Plugin Admin because the directory structure of the ZIP file is pretty plain - just the DLL.  I should look at how other authors of more complex plugins work with Plugin Admin (I think Python Script and XML Tools needed you to copy some DLLs to the Notepad++ "root" directory a while back - maybe that's fixed now).

In any case, going forward I was thinking, the only 2 items in the CONF file are allowedChars which is basically the N++ allowed ANSI chars and languages which again is the N++ language list.  Changing either of these would impact how the plugin works probably for the worse so I wonder why they are even "exposed" to a config file rather than just "hardcoded" or obtained from N++ and not user-configurable.  If that's the case and I do away with them being user-configurable, I don't need that config file anymore.  I just need the database file.

I think plugin config files should be in the NppPluginsDir/config directory.  That's why I was putting QuickText.conf.ini there.  The expansion database I was keeping with the DLL, but realistically, that could be in NppPluginsDir/config like for example `npes_saved.txt` which is the file of all the NppExec scripts if you have that installed.  Of course there's the dilemma of first-time installation vs. upgrade.

First-time:
Copy the default database QuickText.ini to where ever it's suppose to live.  Now the user can start "batteries included".

Upgrade:
Suppose I've made customizations to my QuickText.ini database, I surely don't want an upgrade copying over my existing QuickText.ini and destroying my updates with the original default file and I'm not sure Plugin Admin has logic to detect that. 

I always found the 2 config files confusing myself and not sure how to address so I just left them alone.  I think completely removing the QuickText.conf.ini file wouldn't hurt and then it's just about where to put the database QuickText.ini file.  Probably should be in NppPluginsDir/config instead of with the DLL, but I"m not sure how to make Plugin Admin do that during installation.

Cheers.

**Comment by pryrt** [comment]
I agree, there's not much point to the QuickText.conf.ini, so that could be eliminated.

For how to automatically populate the NppPluginsDir\config, I'm not sure (since I've never released a plugin).  I looked at a handful of plugins: neither [XMLTools](https://github.com/morbac/xmltools/releases/latest) nor [NppFTP](https://github.com/ashkulz/NppFTP/releases/latest), which both put their config files in NppPluginsDir/config (or a subdirectory) contain any of their option files in the zip.  [NppExec](https://github.com/d0vgan/nppexec/releases/latest) also doesn't have any of its config files in the zip, but in the PluginsAdmin-compatible zips (_PA.zip), it does contain a docs folder, which ends up in the installed/plugins/docs directory.  

So, I _think_ if you at least made the zip structure of 
```
zip
    QuickText\QuickText.dll
    Config\QuickText.ini
```
That PluginsAdmin would _probably_ put the QuickText.ini in installdir\plugins\Config\QuickText.ini ... though it _might_ be nice enough to put the Config hierarchy in NppPluginsDir\Config instead.

However, since those other plugins all seem to auto-create the default .ini file, rather than keeping it as text, maybe just autogenerate the .ini in the appropriate NppPluginsDir from inside the plugin, if the file doesn't already exist.  Or maybe take a page from Don's book, and bundle QuickText.ini.model in the DLL directory, and if the NppPluginsDir\Config\QuickText.ini doesn't exist, then copy the file from the .model to the right location.

I think the latter is probably how [FilterHTML](https://fossil.2of4.net/npp_preview/home) does it, because their zip just has the DLL, the release notes, and the filters.sample.ini in the root folder; all of those are in the install\plugins\FilterHTML folder as well; after the plugin has been used, the NppPluginsDir\Config\PreviewHTML subfolder contains filters.ini, settings.ini, and filters.sample.ini (though I might have manually copied that last one at some point, I cannot tell).

I should make clear: this is not urgent, and I am not demanding immediate (or even soonish) fix; I can live with running as administrator when I want to add new shortcuts; I just thought the issue should be documented.

**Comment by vinsworldcom** [comment]
@pryrt 
I thought of that too, keeping the QuickText.default.ini database in the DLL folder and having the DLL check on startup if NppPluginDir/config/QuickText.ini exists.  If not, copy the default (basically assuming it's a first time installation / run) to the proper NppPluginDir/config/QuickText.ini location.  This way any modifications made to the NppPluginDir/config/QuickText.ini would never be overwritten as long as that file exists.

Thanks for not saying it's urgent - the current lockdown situation has my schedule all mixed up - at least I eliminate my long commute now working remotely.  I wonder if you could just do some Windows ACL manipulation to make the QuickText.ini file owned by you so you could run as a normal user and edit it without having to run the whole N++ as admin?  I guess that may be dependent on the database file being in the NppPluginDir/config/QuickText.ini location we previously talked about though?

Cheers.

**Comment by vinsworldcom** [comment]
@pryrt 
If I put up new source code commits are you able to clone and compile yourself, or would you prefer I create a "pre-release" compiled version and put that under the [Releases](https://github.com/vinsworldcom/nppQuickText/releases) tab on this repo?

Cheers.

**Comment by pryrt** [comment]
> are you able to clone and compile yourself
Sorry, don't have a plugin-compatible compile environment yet.  That's a goal for a few months down the road.

> ACL manipulation to make the QuickText.ini file owned by you
That worked. I hadn't remembered whether Windows allowed you to change ACL in the Program Files hierarchy enough that it would truly behave as if normal user had full R/W access.

**Comment by vinsworldcom** [comment]
@pryrt 

Please try this [pre-release](https://github.com/vinsworldcom/nppQuickText/releases/tag/0.1.3.2) when you have a chance.

Cheers.

**Comment by pryrt** [comment]
That seems to do it.
If there isn't a NppPluginDirectory\QuickText.conf.ini, it creates it there when the plugin is loaded.  If I don't have NppPluginDirectory\QuickText.ini, it creates a blank one when I run the Open Tags File menu entry.

And I verified with the portable that it creates and/or uses the ones in installDir\plugins\config.

Other than populating the created QuickText.ini from a master copy, that meets my needs.

**Comment by vinsworldcom** [comment]
> Other than populating the created QuickText.ini from a master copy

I'm working on that.  I'll post here another pre-release when ready if you don't mind testing again.

Cheers.

**Comment by vinsworldcom** [comment]
@pryrt 

New [pre-release](https://github.com/vinsworldcom/nppQuickText/releases/tag/0.1.3.3) published to address moving the default DB (QuickText.default.ini) from Plugin DLL directory to NppPluginDir/config if QuickText.ini does not exist in the config directory.

**First time install:**
```
NppPluginDir/QuickText/QuickText.dll
NppPluginDir/QuickText/QuickText.default.ini
NppPluginDir/config/
```

The QuickText.ini won't exist in config so logic goes it's not there and the QuickText.defaul.ini is so move it and rename:

```
NppPluginDir/QuickText/QuickText.dll
NppPluginDir/config/QuickText.ini
```

**Upgrade:**
Presumed starting state is:

```
NppPluginDir/QuickText/QuickText.dll
NppPluginDir/config/QuickText.ini
```

After the upgrade:

```
NppPluginDir/QuickText/QuickText.dll
NppPluginDir/QuickText/QuickText.default.ini
NppPluginDir/config/QuickText.ini
```

QuickText.default.ini is **not** moved to the config directory since QuickText.ini already exists.

**Unexpected state:**
If for some reason config/QuickText.ini does not exist and the QuickText.default.ini does not exist to move, it just fails silently (no error) and when you open the QuickText => Options dialog, you'll have the language menu, but no tags.  Once you add a tag, it will "autocreate" the config/QuickText.ini file as your database (much the same way QuickText.conf.ini is auto created - which now just contains the allowedChars variable).

**Question:**
I debated moving the QuickText.default.ini or copying it.  Also thought about deleting it from the DLL directory after the first startup since the logic doesn't care it if doesn't find it.  I'm trying to avoid confusion about where the file should live.  Taking your "Don Ho" recommendation for things like stylers.model.xml and langs.model.xml, those "model" files are copied to their non-model counterparts and they continue to exist in the directory.  Maybe I should copy and not move?

Thoughts?

Cheers.

**Comment by pryrt** [comment]
I would vote to copy, not move.  Two reasons: first, the move fails if I am not administrator at the time (or haven't added ACL for my user).  Second, because if the NppPluginDir copy gets deleted at some point and the default version no longer exists, there's nothing to re-create it from.

When I _do_ have ACL permissions, I am able to auto-populate the NppPluginDir file without difficulty.

**Comment by vinsworldcom** [comment]
Agree.  Changed and tested.

So long as you're OK with it, I'll close this issue and publish the new release on master as 0.1.4.1.

**Comment by vinsworldcom** [comment]
Closed: [0.1.4.1](https://github.com/vinsworldcom/nppQuickText/releases/tag/0.1.4.1)
